experimental/packages/opentelemetry-instrumentation-http/src/http.ts

```diff
@@ -479,54 +478,13 @@ export class HttpInstrumentation extends InstrumentationBase<Http> {
 
           instrumentation._headerCapture.server.captureRequestHeaders(span, header => request.headers[header]);
 
-          // Wraps end (inspired by:
-          // https://github.com/GoogleCloudPlatform/cloud-trace-nodejs/blob/master/src/instrumentations/instrumentation-connect.ts#L75)
-          const originalEnd = response.end;
-          response.end = function (
-            this: http.ServerResponse,
-            ..._args: ResponseEndArgs
-          ) {
-            response.end = originalEnd;
-            // Cannot pass args of type ResponseEndArgs,
-            const returned = safeExecuteInTheMiddle(
-              () => response.end.apply(this, arguments as never),
-              error => {
-                if (error) {
-                  utils.setSpanWithError(span, error);
-                  instrumentation._closeHttpSpan(span, SpanKind.SERVER, startTime, metricAttributes);
-                  throw error;
-                }
-              }
-            );
-
-            const attributes = utils.getIncomingRequestAttributesOnResponse(
-              request,
-              response
-            );
-            metricAttributes = Object.assign(metricAttributes, utils.getIncomingRequestMetricAttributesOnResponse(attributes));
-
-            instrumentation._headerCapture.server.captureResponseHeaders(span, header => response.getHeader(header));
-
-            span
-              .setAttributes(attributes)
-              .setStatus({ code: utils.parseResponseStatus(SpanKind.SERVER, response.statusCode) });
-
-            if (instrumentation._getConfig().applyCustomAttributesOnSpan) {
-              safeExecuteInTheMiddle(
-                () =>
-                  instrumentation._getConfig().applyCustomAttributesOnSpan!(
-                    span,
-                    request,
-                    response
-                  ),
-                () => { },
-                true
-              );
-            }
-
-            instrumentation._closeHttpSpan(span, SpanKind.SERVER, startTime, metricAttributes);
-            return returned;
-          };
+          // After 'error', no further events other than 'close' should be emitted.
+          response.on('finish', () => {
+            instrumentation._onServerResponseFinish(request, response, span, metricAttributes, startTime);
+          });
+          response.on(errorMonitor, (err: Err) => {
```

Comment:
I guess we can get rid of the response.end wrapping because we have access to errorMonitor now?

Reply:
I don't think they are related. The problem here is that `response.end` is asynchronous and it can accept an optional callback. We need to close the span when the write is complete rather than at the time when `response.end` is called synchronously.

Here the `errorMonitor` listener is added to make it aligned with the HTTP client instrumentation.